server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -374,11 +377,19 @@ private String getChecksumOfLocalFile(String file) throws IOException {
         return localSegmentChecksumMap.get(file);
     }
 
-    private void deleteStaleCommits() {
+    private void deleteStaleCommitsAsync(String threadpoolName, Runnable onCompletion) {
         try {
-            remoteDirectory.deleteStaleSegments(LAST_N_METADATA_FILES_TO_KEEP);
-        } catch (IOException e) {
-            logger.info("Exception while deleting stale commits from remote segment store, will retry delete post next commit", e);
+            indexShard.getThreadPool().executor(threadpoolName).execute(() -> {
+                try {
+                    remoteDirectory.deleteStaleSegments(LAST_N_METADATA_FILES_TO_KEEP);
+                } catch (IOException e) {
+                    logger.info("Exception while deleting stale commits from remote segment store, will retry delete post next commit", e);
+                }
+            });
+        } catch (Exception e) {
+            logger.info("Exception occurred while scheduling deleteStaleCommits", e);
+        } finally {
+            onCompletion.run();
```

Comment:
Would `onCompletion.run();` not get executed even before the runnable starts to execute most of the times?
I think we can have onCompletion.run() in runnable and in the catch block. That should be sufficient.